CRM/Civicase/Hook/BuildForm/LimitRecipientFieldsToOnlySelectedContacts.php

```diff
@@ -28,10 +28,18 @@ public function run(CRM_Core_Form &$form, $formName) {
    */
   public function limitRecipientFields(CRM_Core_Form &$form) {
     $contacts = array_map(function ($contact) {
```

Comment:
Have we checked if the email_id is actually present in the `$form->_contactDetails` array?
Also we can improve this a bit more since the API call will be made for each contact with array_map. We can move the API call up so it is executed for all contacts once. Also let's specify a return Id so only what we need is returned or else the API call will end up joining to custom tables which we don't need.

Same applies for changes made in `RestrictCaseEmailContacts.php`

```php
     $contactResponse = civicrm_api3('Contact', 'get', [
        'sequential' => 1,
        'return' => ['email_id],
        'id' => ['IN' => array_column($form->_contactDetails, 'contact_id)];
        'options' => ['limit' => 1],
      ]);

    $contactEmailds = array_column( $contactResponse['values'], 'email_id', 'id');

   $contacts = array_map(function ($contact) {
      return [
        'email' => $contact['email'],
        'value' => $contact['contact_id'] . '::' . $contact['email'],
        'display_name' => $contact['display_name'],
        'contact_id' => $contact['contact_id'],
        'email' => CRM_Utils_Array::first($contactEmailds[$contact['contact_id']]);
      ];
    }, $form->_contactDetails);

```

Also wondering if we are particular about the email of the contact we want it sent to whether the primary email instead of picking the first one? Probably something to raise with PM and fix later if changes are required.

NB. Didn't test the code.